File: src/main/java/com/flexpoker/game/query/repository/impl/InMemoryOpenGameForPlayerRepository.java
package com.flexpoker.game.query.repository.impl;

import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import org.springframework.stereotype.Repository;

import com.flexpoker.game.query.dto.GameStage;
import com.flexpoker.game.query.dto.OpenGameForUser;
import com.flexpoker.game.query.repository.OpenGameForPlayerRepository;

@Repository
public class InMemoryOpenGameForPlayerRepository implements OpenGameForPlayerRepository {

    private final Map<UUID, Map<UUID, OpenGameForUser>> openGameForUserMap;

    public InMemoryOpenGameForPlayerRepository() {
        openGameForUserMap = new ConcurrentHashMap<>();
    }

    @Override
    public List<OpenGameForUser> fetchAllOpenGamesForPlayer(UUID playerId) {
        return openGameForUserMap.getOrDefault(playerId, Collections.emptyMap()).values().stream()
                .sorted(Comparator.comparingInt(OpenGameForUser::getOrdinal)).collect(Collectors.toList());
    }

    @Override
    public void deleteOpenGameForPlayer(UUID playerId, UUID gameId) {
        openGameForUserMap.get(playerId).remove(gameId);
    }

    @Override
    public void addOpenGameForUser(UUID playerId, UUID gameId, String gameName) {
        openGameForUserMap.putIfAbsent(playerId, new HashMap<>());
        var openGameForUser = new OpenGameForUser(gameId, null, gameName, GameStage.REGISTERING,
                openGameForUserMap.get(playerId).size(), Collections.emptyList());
        openGameForUserMap.get(playerId).put(gameId, openGameForUser);
    }

    @Override
    public void changeGameStage(UUID playerId, UUID gameId, GameStage gameStage) {
        var openGameForUser = openGameForUserMap.get(playerId).get(gameId);
        var updatedOpenGameForUser = new OpenGameForUser(openGameForUser.getGameId(),
                openGameForUser.getMyTableId(), openGameForUser.getName(), gameStage, openGameForUser.getOrdinal(),
                openGameForUser.getViewingTables());
        openGameForUserMap.get(playerId).put(gameId, updatedOpenGameForUser);
    }

    @Override
    public void assignTableToOpenGame(UUID playerId, UUID gameId, UUID tableId) {
        var openGameForUser = openGameForUserMap.get(playerId).get(gameId);
        var updatedOpenGameForUser = new OpenGameForUser(openGameForUser.getGameId(), tableId,
                openGameForUser.getName(), openGameForUser.getGameStage(), openGameForUser.getOrdinal(),
                openGameForUser.getViewingTables());
        openGameForUserMap.get(playerId).put(gameId, updatedOpenGameForUser);
    }

}

File: src/main/kotlin/com/flexpoker/game/query/repository/impl/InMemoryOpenGameForPlayerRepository.kt
package com.flexpoker.game.query.repository.impl

import com.flexpoker.config.ProfileNames
import com.flexpoker.game.command.GameStage
import com.flexpoker.game.query.dto.OpenGameForUser
import com.flexpoker.game.query.repository.OpenGameForPlayerRepository
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Repository
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors

@Profile(ProfileNames.DEFAULT, ProfileNames.GAME_QUERY_INMEMORY)
@Repository
class InMemoryOpenGameForPlayerRepository : OpenGameForPlayerRepository {

    private val openGameForUserMap: MutableMap<UUID?, MutableMap<UUID?, OpenGameForUser>> = ConcurrentHashMap()

    override fun fetchAllOpenGamesForPlayer(playerId: UUID): List<OpenGameForUser> {
        return openGameForUserMap.getOrDefault(playerId, emptyMap())
            .values
            .stream()
            .sorted(Comparator.comparingInt(OpenGameForUser::ordinal))
            .collect(Collectors.toList())
    }

    override fun deleteOpenGameForPlayer(playerId: UUID, gameId: UUID) {
        openGameForUserMap[playerId]!!.remove(gameId)
    }

    override fun addOpenGameForUser(playerId: UUID, gameId: UUID, gameName: String) {
        openGameForUserMap.putIfAbsent(playerId, HashMap())
        val openGameForUser = OpenGameForUser(gameId, null, gameName, GameStage.REGISTERING,
            openGameForUserMap[playerId]!!.size, emptyList())
        openGameForUserMap[playerId]!![gameId] = openGameForUser
    }

    override fun changeGameStage(playerId: UUID, gameId: UUID, gameStage: GameStage) {
        val openGameForUser = openGameForUserMap[playerId]!![gameId]!!
        val updatedOpenGameForUser = openGameForUser.copy(gameStage = gameStage)
        openGameForUserMap[playerId]!![gameId] = updatedOpenGameForUser
    }

    override fun assignTableToOpenGame(playerId: UUID, gameId: UUID, tableId: UUID) {
        val openGameForUser = openGameForUserMap[playerId]!![gameId]!!
        val updatedOpenGameForUser = openGameForUser.copy(myTableId = tableId)
        openGameForUserMap[playerId]!![gameId] = updatedOpenGameForUser
    }

}